src/backend/distributed/executor/multi_utility.c

```diff
@@ -312,12 +313,15 @@ multi_ProcessUtility(PlannedStmt *pstmt,
 
 	if (IsA(parsetree, CopyStmt))
 	{
-		/* copy parse tree since we might scribble on it to fix the schema name */
-		parsetree = copyObject(parsetree);
+		MemoryContext planCtx = GetMemoryChunkContext(parsetree);
+		MemoryContext previousContext = MemoryContextSwitchTo(planCtx);
 
+		parsetree = copyObject(parsetree);
```

Comment:
I'm pretty hesitant to say we should perform all of `ProcessCopyStmt` under the `parsetree`'s planner context, it seems like a good way to get a memory leak (I don't mean a long-lived one, I just mean running the same function over and over in the same connection could keep adding things to the cached plan's context, right?)

I have two alternatives:

  * Copy the object into your own context, pass it to `ProcessCopyStmt`, then copy the return value into the plan context. This way we're only copying what we end up needing for the cached plan (i.e. the parsetree)
  * Keep the code as-is, except switch back to the previous context before calling `ProcessCopyStmt`. To do this, you'll need to go into `ProcessCopyStmt` and find anywhere you write to the parsetree. So far as I can see, this is its `relation->schemaname` (already handled on line 826 in the past, apparently by me), the filename (on line 913), and the entirety of lines 845 to 870 (which build an entirely new copy statement and select query, which would need to be in the right context)

Performing two copies seems silly but I suppose it's the easier solution?

Reply:
Yeh, this definitely leaks a bunch of memory into `CachedPlan`, which has session lifetime.